crates/wysiwyg/src/composer_model/format.rs

```diff
@@ -51,7 +53,189 @@ where
+        let format = InlineFormatType::InlineCode;
+
+        if s == e {
+            self.toggle_zero_length_format(&format);
+            ComposerUpdate::update_menu_state(
+                self.compute_menu_state(MenuStateComputeType::KeepIfUnchanged),
+            )
+        } else {
+            let range = self.state.dom.find_range(s, e);
+            let leaves = range.leaves();
+            // We'll iterate through the leaves finding their closest structural node ancestor and
+            // grouping these leaves based on the handles of these ancestors.
+            let mut structure_ancestors = HashMap::new();
+            for leaf in leaves {
+                let first_structure_ancestor =
+                    self.find_structure_ancestor(&leaf.node_handle);
+                // Get the closest ancestor path or the root one (empty Vec) if there is none
+                let ancestor_handle = first_structure_ancestor
+                    .map(|a| a.raw().clone())
+                    .unwrap_or(Vec::new());
+                let list = structure_ancestors
+                    .entry(ancestor_handle)
+                    .or_insert(Vec::new());
+                // Add the DomHandle of the leaf to the list of grouped handles by this ancestor
+                list.push(leaf.clone());
+            }
+
+            // Order those ancestors (important to avoid node replacement & conflicts of handles)
+            let mut keys: Vec<&Vec<usize>> =
+                structure_ancestors.keys().collect();
+            keys.sort();
+
+            // Iterate through them backwards, replacing their descendant leaves as needed
+            for ancestor_handle in keys.into_iter().rev() {
+                let leaves = structure_ancestors.get(ancestor_handle).unwrap();
+                // We'll store the text contents of the removed formatted nodes here
+                let mut cur_text = S::default();
+                // Where we'll insert the result of merging the text contents
+                let mut insert_text_at: Option<DomHandle> = None;
+                // Nodes to be added to the Dom, might contain both TextNodes and LineBreaks
+                let mut nodes_to_add = Vec::new();
+                // Iterate the leaves backwards to avoid modifying the previous Dom structure
+                for leaf in leaves.into_iter().rev() {
+                    // Find the immediate child of the common ancestor containing this leaf as its descendant
+                    let (path_to_child, _) = leaf
+                        .node_handle
+                        .raw()
+                        .split_at(ancestor_handle.len() + 1)
+                        .clone();
+                    let ancestor_child =
+                        DomHandle::from_raw(path_to_child.to_vec());
+
+                    let node =
+                        self.state.dom.lookup_node(&leaf.node_handle).clone();
+                    match node {
```

Comment:
Can we break out the arms of the match into methods, or similar? There is a lot of code here.

Reply:
Yeah, that sounds good.